<!DOCTYPE html>
<html lang='en'>
<head>
<title>0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch\trunk - svntogit/packages.git - Git clone of the 'packages' repository
</title>
<meta name='generator' content='cgit v1.2.1-1-g437b'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit.css'/>
<link rel='shortcut icon' href='/favicon.ico'/>
<link rel='alternate' title='Atom feed' href='https://git.archlinux.org/svntogit/packages.git/atom/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils' type='application/atom+xml'/>
<link rel='vcs-git' href='https://git.archlinux.org/svntogit/packages.git' title='svntogit/packages.git Git repository'/>
<link rel='vcs-git' href='git://git.archlinux.org/svntogit/packages.git' title='svntogit/packages.git Git repository'/>
<link rel='vcs-git' href='ssh://git.archlinux.org/srv/git/svntogit/packages.git' title='svntogit/packages.git Git repository'/>
</head>
<body>
	<div id="archnavbar"><!-- Arch Linux global navigation bar -->
		<div id="archnavbarlogo">
			<p><a href="http://www.archlinux.org/" title="Arch news, packages, projects and more"></a></p>
		</div>
		<div id="archnavbarmenu">
			<ul id="archnavbarlist">
				<li id="anb-home"><a href="http://www.archlinux.org/" title="Arch news, packages, projects and more">Home</a></li>
				<li id="anb-packages"><a href="http://www.archlinux.org/packages/" title="Arch Package Database">Packages</a></li>
				<li id="anb-forums"><a href="https://bbs.archlinux.org/" title="Community forums">Forums</a></li>
				<li id="anb-wiki"><a href="https://wiki.archlinux.org/" title="Community documentation">Wiki</a></li>
				<li id="anb-bugs"><a href="https://bugs.archlinux.org/" title="Report and follow bugs">Bugs</a></li>
				<li id="anb-sec"><a href="https://security.archlinux.org/" title="Security Tracker">Security</a></li>
				<li id="anb-aur"><a href="https://aur.archlinux.org/" title="Arch Linux User Repository">AUR</a></li>
				<li id="anb-download"><a href="http://www.archlinux.org/download/" title="Get Arch Linux">Download</a></li>
			</ul>
		</div>
	</div><!-- #archnavbar -->
<div id='cgit'><table id='header'>
<tr>
<td class='main'><a href='/'>index</a> : <a title='svntogit/packages.git' href='/svntogit/packages.git/'>svntogit/packages.git</a></td></tr>
<tr><td class='sub'>Git clone of the 'packages' repository
</td><td class='sub right'></td></tr></table>
<table class='tabs'><tr><td>
<a href='/svntogit/packages.git/?h=packages/binutils'>summary</a><a href='/svntogit/packages.git/refs/?h=packages/binutils'>refs</a><a href='/svntogit/packages.git/log/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>log</a><a class='active' href='/svntogit/packages.git/tree/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>tree</a><a href='/svntogit/packages.git/commit/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>commit</a><a href='/svntogit/packages.git/diff/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>diff</a><a href='/svntogit/packages.git/stats/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>stats</a></td><td class='form'><form class='right' method='get' action='/svntogit/packages.git/log/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch'>
<input type='hidden' name='h' value='packages/binutils'/><select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='search' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/svntogit/packages.git/tree/?h=packages/binutils'>root</a>/<a href='/svntogit/packages.git/tree/trunk?h=packages/binutils'>trunk</a>/<a href='/svntogit/packages.git/tree/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch</a></div><div class='content'>blob: d56d8b7b6fd2cc0409883fae35d1c2e146732975 (<a href='/svntogit/packages.git/plain/trunk/0004-PR23486-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED-x86_64.patch?h=packages/binutils'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
<a id='n32' href='#n32'>32</a>
<a id='n33' href='#n33'>33</a>
<a id='n34' href='#n34'>34</a>
<a id='n35' href='#n35'>35</a>
<a id='n36' href='#n36'>36</a>
<a id='n37' href='#n37'>37</a>
<a id='n38' href='#n38'>38</a>
<a id='n39' href='#n39'>39</a>
<a id='n40' href='#n40'>40</a>
<a id='n41' href='#n41'>41</a>
<a id='n42' href='#n42'>42</a>
<a id='n43' href='#n43'>43</a>
<a id='n44' href='#n44'>44</a>
<a id='n45' href='#n45'>45</a>
<a id='n46' href='#n46'>46</a>
<a id='n47' href='#n47'>47</a>
<a id='n48' href='#n48'>48</a>
<a id='n49' href='#n49'>49</a>
<a id='n50' href='#n50'>50</a>
<a id='n51' href='#n51'>51</a>
<a id='n52' href='#n52'>52</a>
<a id='n53' href='#n53'>53</a>
<a id='n54' href='#n54'>54</a>
<a id='n55' href='#n55'>55</a>
<a id='n56' href='#n56'>56</a>
<a id='n57' href='#n57'>57</a>
<a id='n58' href='#n58'>58</a>
<a id='n59' href='#n59'>59</a>
<a id='n60' href='#n60'>60</a>
<a id='n61' href='#n61'>61</a>
<a id='n62' href='#n62'>62</a>
<a id='n63' href='#n63'>63</a>
<a id='n64' href='#n64'>64</a>
<a id='n65' href='#n65'>65</a>
<a id='n66' href='#n66'>66</a>
<a id='n67' href='#n67'>67</a>
<a id='n68' href='#n68'>68</a>
<a id='n69' href='#n69'>69</a>
<a id='n70' href='#n70'>70</a>
<a id='n71' href='#n71'>71</a>
<a id='n72' href='#n72'>72</a>
<a id='n73' href='#n73'>73</a>
<a id='n74' href='#n74'>74</a>
<a id='n75' href='#n75'>75</a>
<a id='n76' href='#n76'>76</a>
<a id='n77' href='#n77'>77</a>
<a id='n78' href='#n78'>78</a>
</pre></td>
<td class='lines'><pre><code>From 96027b49650718ebcd20d81bd8d0fbbd36a33558 Mon Sep 17 00:00:00 2001
From: Cary Coutant &lt;ccoutant&#64;gmail.com&gt;
Date: Tue, 7 Aug 2018 21:35:12 -0700
Subject: [PATCH] Properly merge GNU_PROPERTY_X86_ISA_1_USED (x86_64).

gold/
	PR ld/23486
	* x86_64.cc (Target_x86_64::Target_x86_64): Initialize
	object_isa_1_used_.
	(Target_x86_64::object_isa_1_used_): New data member.
	(Target_x86_64::record_gnu_property): Save ISA_1_USED bits for object.
	(Target_x86_64::merge_gnu_properties): Merge ISA_1_USED bits.
<span class="hl kwb">---</span>
 gold/x86_64.cc | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/gold/x86_64.cc b/gold/x86_64.cc
index 27f273d64b..ac958616c6 100644
<span class="hl kwb">--- a/gold/x86_64.cc</span>
<span class="hl kwa">+++ b/gold/x86_64.cc</span>
&#64;&#64; -706,7 +706,8 &#64;&#64; class Target_x86_64 : public Sized_target&lt;size, false&gt;
       rela_irelative_(NULL), copy_relocs_(elfcpp::R_X86_64_COPY),
       got_mod_index_offset_(-1U), tlsdesc_reloc_info_(),
       tls_base_symbol_defined_(false), isa_1_used_(0), isa_1_needed_(0),
<span class="hl kwb">-      feature_1_(0), object_feature_1_(0), seen_first_object_(false)</span>
<span class="hl kwa">+      feature_1_(0), object_isa_1_used_(0), object_feature_1_(0),</span>
<span class="hl kwa">+      seen_first_object_(false)</span>
   { }
 
   // Hook for a new output section.
&#64;&#64; -1381,6 +1382,11 &#64;&#64; class Target_x86_64 : public Sized_target&lt;size, false&gt;
   uint32_t isa_1_needed_;
   uint32_t feature_1_;
   // Target-specific properties from the current object.
<span class="hl kwa">+  // These bits get ORed into ISA_1_USED_ after all properties for the object</span>
<span class="hl kwa">+  // have been processed. But if either is all zeroes (as when the property</span>
<span class="hl kwa">+  // is absent from an object), the result should be all zeroes.</span>
<span class="hl kwa">+  // (See PR ld/23486.)</span>
<span class="hl kwa">+  uint32_t object_isa_1_used_;</span>
   // These bits get ANDed into FEATURE_1_ after all properties for the object
   // have been processed.
   uint32_t object_feature_1_;
&#64;&#64; -1609,7 +1615,7 &#64;&#64; Target_x86_64&lt;size&gt;::record_gnu_property(
   switch (pr_type)
     {
     case elfcpp::GNU_PROPERTY_X86_ISA_1_USED:
<span class="hl kwb">-      this-&gt;isa_1_used_ |= val;</span>
<span class="hl kwa">+      this-&gt;object_isa_1_used_ |= val;</span>
       break;
     case elfcpp::GNU_PROPERTY_X86_ISA_1_NEEDED:
       this-&gt;isa_1_needed_ |= val;
&#64;&#64; -1627,12 +1633,22 &#64;&#64; void
 Target_x86_64&lt;size&gt;::merge_gnu_properties(const Object*)
 {
   if (this-&gt;seen_first_object_)
<span class="hl kwb">-    this-&gt;feature_1_ &amp;= this-&gt;object_feature_1_;</span>
<span class="hl kwa">+    {</span>
<span class="hl kwa">+      // If any object is missing the ISA_1_USED property, we must omit</span>
<span class="hl kwa">+      // it from the output file.</span>
<span class="hl kwa">+      if (this-&gt;object_isa_1_used_ == 0)</span>
<span class="hl kwa">+	this-&gt;isa_1_used_ = 0;</span>
<span class="hl kwa">+      else if (this-&gt;isa_1_used_ != 0)</span>
<span class="hl kwa">+	this-&gt;isa_1_used_ |= this-&gt;object_isa_1_used_;</span>
<span class="hl kwa">+      this-&gt;feature_1_ &amp;= this-&gt;object_feature_1_;</span>
<span class="hl kwa">+    }</span>
   else
     {
<span class="hl kwa">+      this-&gt;isa_1_used_ = this-&gt;object_isa_1_used_;</span>
       this-&gt;feature_1_ = this-&gt;object_feature_1_;
       this-&gt;seen_first_object_ = true;
     }
<span class="hl kwa">+  this-&gt;object_isa_1_used_ = 0;</span>
   this-&gt;object_feature_1_ = 0;
 }
 
<span class="hl kwb">-- </span>
2.18.0

</code></pre></td></tr></table>
</div> <!-- class=content -->
<div class="foot" style="padding-left:1em;padding-right:1em;">
<p>Copyright &copy; 2002-2018 <a href="mailto:jvinet@zeroflux.org"
title="contact Judd Vinet">Judd Vinet</a> and <a href="mailto:aaron@archlinux.org"
title="contact Aaron Griffin">Aaron Griffin</a>. The Arch Linux name and logo
are recognized trademarks. Some rights reserved. The registered trademark
Linux&reg; is used pursuant to a sublicense from LMI, the exclusive licensee
of Linus Torvalds, owner of the mark on a world-wide basis.</p>
</div>
</div> <!-- id=cgit -->
</body>
</html>
